src/s3/commands/uploadFile.ts

```diff
@@ -114,11 +207,167 @@ async function uploadWithProgress({
+                'Failed to list buckets from client, please try changing your region then try again',
+            ),
+            localize('AWS.message.prompt.changeS3RegionButton','Change default S3 region')
+        )
+        .then((selection: string | undefined) => {
+            if (selection === 'Change default S3 region') {
+                vscode.commands.executeCommand('workbench.action.openSettings', 'aws.s3.defaultRegion')
+            } 
+        })
+        telemetry.recordS3UploadObject({ result: 'Failed' })
+        throw new Error('Failed to list buckets from client')
+    }
+    
+    const s3Buckets = allBuckets.filter(bucket => {
+        return bucket && bucket.Name
+    }) as S3.Bucket[]
+
+    
+    const createNewBucket: BucketQuickPickItem = {
+        label: localize('AWS.command.s3.createBucket',"Create new bucket"),
+        bucket: undefined
+    }
+    const bucketItems: BucketQuickPickItem[] = s3Buckets.map(bucket => {
+        return { 
+            label: bucket.Name!,
+            bucket
+        }
+    })
+
+    const picker = createQuickPick({
+        options: {
+            canPickMany: false,
+            ignoreFocusOut: true,
+            title: localize('AWS.message.selectBucket','Select an S3 bucket to upload to'),
+            step: 2,
+            totalSteps: 2,
+        },
+        items: [...bucketItems, createNewBucket],
+        buttons: [vscode.QuickInputButtons.Back]
+    })
+    const response = verifySinglePickerOutput(await promptUserFunction({
+        picker: picker,
+        onDidTriggerButton: (button, resolve, reject) => {
+            if (button === vscode.QuickInputButtons.Back) {
+                resolve([{
+                    label: 'back',
+                    bucket: undefined
+                }])
+            }
+
+        },
+    })) 
+
+    if (!response) {
+        return 'cancel'
+    }
+
+    if (!response.bucket) {
+        if (response.label === 'back') {
+            return response.label
+        }
+        if (response.label === 'Create new bucket') {
+            const s3Node = new S3Node(s3client)
+            await createBucket(s3Node, window, Commands.vscode())
+            return promptUserForBucket(s3client)
+        }
+    } else {
+        return response.bucket
+    }
+    return 'cancel'
+}
+
+/**
+ * Gets the open file in the current editor
+ * Asks the user to browse for more files
+ * If no file is open it prompts the user to select file
+ * @param document document to use as currently open
+ * 
+ * @returns file selected by the user
+ */
+export async function getFileToUpload(
+    document?: vscode.Uri,
+    window = Window.vscode(),
+    promptUserFunction = promptUser
+): Promise<vscode.Uri| undefined> {
+    
+    let fileLocation: vscode.Uri | undefined
+
+    if (!document) {
+        fileLocation = await promptForFileLocation(window)
+    } 
+    else {
+        fileLocation = document
```

Comment:
Nit: Seems a little goofy to set the value before you prompt the user. Is there ever a case where you'd have multiple items in the prompt? Is there a case where you could leave the prompter without selecting but still return this value (e.g. `response` is defined but either doesn't have a label or the label doesn't match the Browse More Files string)

Reply:
There is no case where the prompt will have multiple items. It just gets the document that is currently active on the editor. I don't think that is a possibility, since the only 4 outcomes of this prompt (when it goes to the else at the begging) are: the user presses the back button which will set response to undefined and return that, user exits the quick pick which will return undefined immediately, user selects browse for more files, which will run promptForFileLocation() and return that,  that in fact I just changed the line to be return promptForFileLocation(window), and last, the user selects to use the current open file and it returns that